### PR TITLE
chore: update k8s packages to latest version

### LINKED ISSUE #573: Bump `knative.dev/pkg` and k8s packages to `v0.27.6`
### Description

**What problem are you trying to solve?**

Karpenter is currently bottlenecked on upgrading its dependencies by the upstream `knative.dev/pkg` dependency. `knative.dev/pkg` has updated its latest [`go.mod`](https://github.com/knative/pkg/blob/main/go.mod) to support `v0.27.6` so we should be able to update our k8s client packages to that version. That includes the following bumps:

`knative.dev/pkg`: `v0.0.0-20231003141102-833dd976f13d`
k8s packages (e.g. `k8s.io/client-go` or `k8s.io/apimachinery`): `v0.27.6`
`controller-runtime`: `v0.15.2`

Note: These changes may cause some code changes due to API changes or deprecations.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by sadath-12**
Fixes #573 

**Description**
This change was important as for using features like sidecar we need upto date k8s version

**How was this change tested?**
```make presubmit```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -21,16 +21,16 @@ require (
 	golang.org/x/sync v0.3.0
 	golang.org/x/text v0.13.0
 	golang.org/x/time v0.3.0
-	k8s.io/api v0.26.6
-	k8s.io/apiextensions-apiserver v0.26.6
-	k8s.io/apimachinery v0.26.6
-	k8s.io/client-go v0.26.6
+	k8s.io/api v0.28.2
+	k8s.io/apiextensions-apiserver v0.28.0
```
Comment (jonathan-innis) [inline go.mod:25]: Why is this at a different version? Typically all of the k8s client package versions are able to match

File: go.mod [hunk 2]
```diff
@@ -21,16 +21,16 @@ require (
...
+	k8s.io/apimachinery v0.28.2
+	k8s.io/client-go v0.28.2
 	k8s.io/cloud-provider v0.26.6
 	k8s.io/csi-translation-lib v0.26.6
 	k8s.io/klog/v2 v2.100.1
-	k8s.io/utils v0.0.0-20230209194617-a36077c30491
+	k8s.io/utils v0.0.0-20230406110748-d93618cff8a2
 	knative.dev/pkg v0.0.0-20230712131115-7051d301e7f4
```
Comment (jonathan-innis) [inline go.mod:32]: Interesting that we didn't have to update knative.dev/pkg here. Seems like somehow we are getting away here without bumping the package version.

**Comment by sadath-12** [comment]
Do you want me to try updating the other k8s versions ? Because for init containers I hate to update core pkg which in return asked me to update client-go and controller-runtime

**Comment by sadath-12** [comment]
Wanted to mention there 2 places I commented off since I need a bit of help in transforming them

File: go.mod (hunk 2 above)
Comment (sadath-12) [reply go.mod:32]: yeah hopefully

File: go.mod (hunk 1 above)
Comment (sadath-12) [reply go.mod]: ya done

File: pkg/operator/profiling.go
Comment (sadath-12) [inline pkg/operator/profiling.go:1]: AddMetricsExtraHandler function is no longer supported

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -118,13 +116,13 @@ func NewOperator() (context.Context, *Operator) {
...
 			return ctx
 		},
-		NewCache: cache.BuilderWithOptions(cache.Options{
-			SelectorsByObject: cache.SelectorsByObject{
-				&coordinationv1.Lease{}: {
-					Field: fields.SelectorFromSet(fields.Set{"metadata.namespace": "kube-node-lease"}),
-				},
-			},
-		}),
+		// NewCache: cache.NewCacheFunc(config,cache.Options{
```
Comment (sadath-12) [inline pkg/operator/operator.go:119]: this thing throws some warning if uncommented but not an error

**Comment by sadath-12** [comment]
@jonathan-innis  all k8s packages are updated now

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Review by jonathan-innis — DISMISSED** [review]
LGTM 🚀

**Review by njtran — APPROVED** [review]

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Pull request merged**